Created understanding gitops concept module

jc-berger opened the pull request:
- Adds an overview for GitOps and addresses Jira Issue, [#2475](https://issues.redhat.com/browse/RHDEVDOCS-2475?filter=-1).

- Targeted to OCP 4.7

- Aligned team label: Dev Tools

- Acked by SME/QE

- For easy reference, here's a direct preview link: https://deploy-preview-29316--osdocs.netlify.app/openshift-enterprise/latest/cicd/gitops/understanding-openshift-gitops.html

Thanks, all!

dewan-ahmed commented:
Thanks @jc-berger for this PR. Are we referencing the 'getting-started' doc somewhere here too?

jc-berger commented:
Hey @dewan-ahmed, thanks for asking. We do intend to have a `getting-started` doc, but not included in this `understanding gitops` doc. Once I get a getting started doc made in a separate PR, I can reference it in this overview doc, but we want them as separate docs. Because, this overview, `understanding gitops` doc is concept based, and the `getting started` doc is procedure/action based.

jc-berger commented on modules/gitops-document-attributes.adoc:
```diff
@@ -0,0 +1,12 @@
+//
+:gitops-title: Red Hat OpenShift GitOps
+:gitops-shortname: GitOps
+:gitops-ver: release-tech-preview-4.7
```
confirm what this version is called. Thinking it would be 1.0 since it's the first GitOps version.

jc-berger commented on modules/openshift-argocd-concept.adoc:
```diff
@@ -0,0 +1,12 @@
+[id="openshift-argocd-concept"]
```
Title and ID must be as similar :)

jc-berger commented on modules/understanding-openshift-gitops.adoc:
```diff
@@ -0,0 +1,39 @@
+[id="understanding-openshift-gitops"]
+= Understanding OpenShift GitOps
```
Consider calling it "About..." or "Overview..." "Understanding..." is similar to "Working...." which will be the name of our assembly.

Preeticp commented on modules/overview-of-openshift-gitops.adoc:
```diff
@@ -0,0 +1,39 @@
+
+Use Red Hat OpenShift GitOps to create repeatable processes for managing OpenShift Container Platform clusters and applications across multi-cluster Kubernetes environments. When you lose time in deployment and release cycles, Red Hat Openshift GitOps is a tool that handles and automates the fast pace and complex deployments. The Red Hat Openshift GitOps’ workflow pushes from application tp devel ...
+
+Red Hat Openshift GitOps is a declarative way to implement continuous deployment for cloud knative applications. Red Hat Openshift GitOps has a Git repository that contains a declarative description of the infrastructure you need in your specified environment. This Git repository also contains an automated process to make your environment match the described state. Basically,  Red Hat Openshift G ...
```
This entire paragraph complements other sections that are explained in a much better manner in the section below.
I have made suggestions on where they can add more value.
Remove the word 'Basically' and the exclamation mark at the end.

Preeticp commented on modules/overview-of-openshift-gitops.adoc:
```diff
@@ -0,0 +1,39 @@
+Red Hat Openshift GitOps organizes the deployment process around the code repositories and makes them the central element. Red Hat Openshift GitOps always has at least two repositories:
+
+  . Application repository with the source code
+  . Environment configuration repository that defines the desired state of the application
```
These sentences add more context here, suggest adding them here. Something like - 
These (please confirm which of the above repositories) repositories contain a declarative description of the infrastructure you need in your specified environment. It/They(?) also contain an automated process to make your environment match the described state.

Preeticp commented on modules/overview-of-openshift-gitops.adoc:
```diff
@@ -0,0 +1,39 @@
+:FeatureName: OpenShift GitOps
+include::modules/technology-preview.adoc[leveloffset=+1]
+
+Use Red Hat OpenShift GitOps to create repeatable processes for managing OpenShift Container Platform clusters and applications across multi-cluster Kubernetes environments. When you lose time in deployment and release cycles, Red Hat Openshift GitOps is a tool that handles and automates the fast pace and complex deployments. The Red Hat Openshift GitOps’ workflow pushes from application tp devel ...
```
Suggest adding this sentence here: It either deploys a new application or updates an existing one, so that you only need to update the repository.  Red Hat Openshift GitOps handles and automates everything else.

Preeticp commented on modules/overview-of-openshift-gitops.adoc:
```diff
@@ -0,0 +1,39 @@
+
+== Additional resources
+
+* For information on Red Hat Openshift GitOps, see the xref:../architecture/cicd_gitops.adoc#cicd-methodology-and-practice[CI/CD methodology and practice].
```
This section has been removed, so remove these links.

Preeticp commented on modules/overview-of-openshift-gitops.adoc:
```diff
@@ -0,0 +1,39 @@
+
+:FeatureName: OpenShift GitOps
+include::modules/technology-preview.adoc[leveloffset=+1]
+
```
Use the first 2 paragraphs here to first explain what is GitOps - Maybe a small module - About GitOps https://docs.openshift.com/container-platform/4.6/architecture/cicd_gitops.html#cicd_gitops_methodology

This sentence could be added at the start of this paragraph on line 11: Red Hat Openshift GitOps is a declarative way to implement continuous deployment for cloud knative applications.

Preeticp commented on modules/openshift-argocd-concept.adoc:
```diff
@@ -0,0 +1,12 @@
+[id="openshift-argocd-concept"]
+= OpenShift ArgoCD
```
Do we really have a OpenShift ArgoCD? If yes, rename to About OpenShift ArgoCD
If not remove this module and move its content to the Overview of OpenShift GitOps section.

Preeticp commented on modules/openshift-argocd-concept.adoc:
```diff
@@ -0,0 +1,12 @@
+:context: openshift-argocd-concept
+
+:FeatureName: ArgoCD
+include::modules/technology-preview.adoc[leveloffset=+1]
```
The entire feature is TP so we do not need to have includes here.
Also we do not add includes to modules, only to assemblies.

Preeticp commented on modules/overview-of-openshift-gitops.adoc:
```diff
@@ -0,0 +1,39 @@
+* Promote applications across clusters, from stage to production
+
+//Main Concept Topic
+include::modules/openshift-argocd-concept.adoc[leveloffset=+1]
```
This include statement needs to be part of the Assembly. Do not add include statements in a module. This is like adding a module to a module.

Preeticp commented on modules/overview-of-openshift-gitops.adoc:
```diff
@@ -0,0 +1,39 @@
+include::modules/common-attributes.adoc[]
+include::modules/gitops-document-attributes.adoc[]
+:context: overview-of-openshift-gitops
+toc::[]
```
Please update the metadata information to match that for modules. Please refer to module templates in the [modular documentation guide](https://redhat-documentation.github.io/modular-docs/#introduction_{context}) to understand how modules are created. Or to a module in the modules folder of the OpenShift docs repository. This uses the template used for Assemblies in the OCP docs repository. The OpenShift specific guidelines are noted here, please ensure that you go through them: https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc

Preeticp replied:
Removing the include statements and correcting the metadata in the modules might help the build succeed.

jc-berger commented:
New, squashed commit is up, addressing Preeti's first round of doc peer review. @Preeticp, please review again. 

@sbose78, I Slacked Shubham (can't seem to find him on GH), but would one of you please do a dev review? 

Thanks, y'all!

Preeticp commented on cicd/gitops/working-with-openshift-gitops.adoc:
```diff
@@ -0,0 +1,18 @@
+include::modules/overview-of-openshift-gitops.adoc[leveloffset=+1]
+
+//Procedure Module
+include::modules/getting-started-openshift-gitops.adoc[leveloffset=+1]
```
Add // at the beginning of this include statement since it is not added in the same PR, or remove it, you can add it in the other PR.

Preeticp commented on modules/about-openshift-gitops.adoc:
```diff
@@ -0,0 +1,10 @@
+
+Red Hat OpenShift GitOps is a declarative way to implement continuous deployment for cloud knative appliactions. Use Red Hat OpenShift GitOps to create repeatable processes for managing OpenShift Container Platform clusters and applications across multi-cluster Kubernetes environments. Red Hat OpenShift GitOps handles and automates complex deployments at a fast pace, saving time during deployment ...
+
+Red Hat OpenShift GitOps is a declarative way to implement continuous deployment for cloud knative applications. Red Hat OpenShift GitOps has a Git repository that contains a declarative description of the infrastructure you need in your specified environment. This Git repository also contains an automated process to make your environment match the described state.
```
Remove: Red Hat OpenShift GitOps is a declarative way to implement continuous deployment for cloud knative applications. since you added it above

Preeticp commented on modules/about-openshift-gitops.adoc:
```diff
@@ -0,0 +1,10 @@
+// * openshift-docs/cicd/gitops/working-with-openshift-gitops.adoc
+
+[id="about-openshift-gitops_{context}"]
+= About OpenShift GitOps
```
s/About OpenShift GitOps/ = About GitOps

There is no difference in About OpenShift GitOps and Overview of OpenShift GitOps. Suggest making the first section - About GitOps and then About Red Hat OpenShift GitOps 

I do not see the include statement for this module in the working-with-openshift-gitops.adoc assembly.

Preeticp commented on modules/overview-of-openshift-gitops.adoc:
```diff
@@ -0,0 +1,32 @@
+// * openshift-docs/cicd/gitops/working-with-openshift-gitops.adoc
+
+[id="overview-of-openshift-gitops_{context}"]
+= Overview of OpenShift GitOps
```
s/Overview of OpenShift GitOps/ Overview of {gitops-title}

Preeticp commented on modules/overview-of-openshift-gitops.adoc:
```diff
@@ -0,0 +1,32 @@
+[id="overview-of-openshift-gitops_{context}"]
+= Overview of OpenShift GitOps
+
+Red Hat OpenShift GitOps is a set of practices that use Git pull requests to manage infrastructure and application configurations. The Git repository in Red Hat OpenShift GitOps is the only source of truth for system and application configuration. The repository contains the entire state of the system so that the trail of changes to the system state are visible and auditable.
```
s/Red Hat OpenShift GitOps/ GitOps througout this paragraph.

The idea is to explain to users what is gitops first and then move on to explain what Red Hat OpenShift GitOps does for you.

Preeticp commented on modules/overview-of-openshift-gitops.adoc:
```diff
@@ -0,0 +1,32 @@
+
+Red Hat OpenShift GitOps is a set of practices that use Git pull requests to manage infrastructure and application configurations. The Git repository in Red Hat OpenShift GitOps is the only source of truth for system and application configuration. The repository contains the entire state of the system so that the trail of changes to the system state are visible and auditable.
+
+Use Red Hat OpenShift GitOps tooling to create repeatable and predictable processes for managing and recreating OpenShift Container Platform clusters and applications. By using Red Hat OpenShift GitOps, you resolve the issues of infrastructure and application configuration sprawl.
```
s/ Use Red Hat OpenShift GitOps tooling/You can use GitOps tooling
s/Red Hat OpenShift GitOps/ GitOps througout this paragraph.

Preeticp commented on modules/overview-of-openshift-gitops.adoc:
```diff
@@ -0,0 +1,32 @@
+
+Use Red Hat OpenShift GitOps tooling to create repeatable and predictable processes for managing and recreating OpenShift Container Platform clusters and applications. By using Red Hat OpenShift GitOps, you resolve the issues of infrastructure and application configuration sprawl.
+
+Red Hat OpenShift GitOps defines infrastructure and application definitions as “code.” Then, Red Hat OpenShift GitOps uses this code to manage multiple workspaces and clusters and to simplify the creation of infrastructure and application configurations. By following the code’s principles, you can store the configuration of clusters and applications in Git repositories then follow the Git workflo ...
```
s/Red Hat OpenShift GitOps/ GitOps througout this paragraph.
"code". period outside the quotation mark.

Make this sentence the last one in this paragraph:
You can apply the core principles of developing and maintaining software in a Git repository to the creation and management of your cluster and application configuration files.

Remove this line from this paragraph and add it to the next:
Red Hat OpenShift GitOps ensures consistency in applications when you deploy them to different clusters in different environments, like: development, stage, and production.

Preeticp commented on cicd/gitops/working-with-openshift-gitops.adoc:
```diff
@@ -0,0 +1,18 @@
+:context: working-with-openshift-gitops
+toc::[]
+
+//Concept Module
```
Add include statement for: About GitOps

Preeticp commented on cicd/gitops/working-with-openshift-gitops.adoc:
```diff
@@ -0,0 +1,21 @@
+toc::[]
+
+//Concept Module
+include::modules/about-openshift-gitops.adoc[leveloffset=+1]
```
Let's change the id to about-gitops.adoc
I would consider Red Hat OpenShift GitOps and OpenShift GitOps synonymous.

Preeticp commented on modules/about-redhat-openshift-gitops.adoc:
```diff
@@ -0,0 +1,26 @@
+
+These repositories contain a declarative description of the infrastructure you need in your specified environment. They also contain an automated process to make your environment match the described state.
+
+{gitops-title} uses ArgoCD to maintain cluster resources. ArgoCD is an open-source declarative tool for the continuous integration and continuous deployment (CI/CD) of applications. Red Hat OpenShift GitOps implements ArgoCD as a controller so that it continuously monitors application definitions and configurations defined in a Git repository. Then, ArgoCD compares the specified state of these co ...
```
s/Red Hat OpenShift GitOps/{gitops-title}

jc-berger commented on _topic_map.yml:
```diff
@@ -1277,16 +1277,12 @@ Topics:
+  Dir: gitops
+  Distros: openshift-enterprise,openshift-webscale,openshift-origin
+  Topics:
+  - Name: Working with OpenShift GitOps
```
Retain the first and third assemblies. Remove "installing openshift..." remove "working with..."
add "getting started..." and "understanding.."

Preeticp commented on _topic_map.yml:
```diff
@@ -1277,16 +1277,14 @@ Topics:
+  Topics:
+  - Name: Understanding OpenShift GitOps
+    File: understanding-openshift-gitops
+  - Name: Getting started with OpenShift GitOps
```
Please comment out this line and the next to resolve the build error.

jc-berger replied:
noted, thanks!

jc-berger commented on cicd/gitops/understanding-openshift-gitops.adoc:
```diff
@@ -0,0 +1,14 @@
+//Concept Module
+include::modules/about-redhat-openshift-gitops.adoc[leveloffset=+1]
+
+[id="openshift-gitops-additional-resources"]
```
I know I need to delete this. I will do so tomorrow!

Preeticp commented on modules/about-gitops.adoc:
```diff
@@ -0,0 +1,12 @@
+[id="about-gitops_{context}"]
+= About GitOps
+
+GitOps is a declarative way to implement continuous deployment for cloud native applications. You can use GitOps to create repeatable processes for managing OpenShift Container Platform clusters and applications across multi-cluster Kubernetes environments. GitOps handles and automates complex deployments at a fast pace, saving time during deployment and release cycles. Its workflow pushes an app ...
```
s/Its workflow/The GitOps workflow

bigkevmcd replied:
You use "staging" and "stage" in this document to refer to the same environment, you could probably standardise on "staging".

Preeticp commented on modules/about-gitops.adoc:
```diff
@@ -0,0 +1,12 @@
+
+GitOps is a declarative way to implement continuous deployment for cloud native applications. You can use GitOps to create repeatable processes for managing OpenShift Container Platform clusters and applications across multi-cluster Kubernetes environments. GitOps handles and automates complex deployments at a fast pace, saving time during deployment and release cycles. Its workflow pushes an app ...
+
+GitOps is a set of practices that use Git pull requests to manage infrastructure and application configurations. Its Git repository is the only source of truth for system and application configuration. This Git repository contains a declarative description of the infrastructure you need in your specified environment and contains an automated process to make your environment match the described st ...
```
s/ Its Git repository / In GitOps, the Git repository

bigkevmcd replied:
Might want to explain how GitOps differs from [Infrastructure-as-Code](https://docs.microsoft.com/en-us/azure/devops/learn/what-is-infrastructure-as-code), which also resolves the issues of infrastructure and application configuration sprawl?

Preeticp commented on modules/about-gitops.adoc:
```diff
@@ -0,0 +1,12 @@
+
+GitOps is a set of practices that use Git pull requests to manage infrastructure and application configurations. Its Git repository is the only source of truth for system and application configuration. This Git repository contains a declarative description of the infrastructure you need in your specified environment and contains an automated process to make your environment match the described st ...
+
+GitOps defines infrastructure and application definitions as “code”. Then, it uses this code to manage multiple workspaces and clusters and to simplify the creation of infrastructure and application configurations. By following the code’s principles, you can store the configuration of clusters and applications in Git repositories then follow the Git workflow to apply these repositories to your ch ...
```
...and clusters and to simplify.../and clusters to simplify
the code’s principles/the principles of code - avoid possessives for inanimate objects 
...Git repositories then follow the Git workflow.../...Git repositories, and then follow...

Preeticp commented on modules/about-redhat-openshift-gitops.adoc:
```diff
@@ -0,0 +1,26 @@
+
+These repositories contain a declarative description of the infrastructure you need in your specified environment. They also contain an automated process to make your environment match the described state.
+
+{gitops-title} uses Argo CD to maintain cluster resources. Argo CD is an open-source declarative tool for the continuous integration and continuous deployment (CI/CD) of applications. Red Hat OpenShift GitOps implements Argo CD as a controller so that it continuously monitors application definitions and configurations defined in a Git repository. Then, Argo CD compares the specified state of thes ...
```
Red Hat OpenShift GitOps/{gitops-title}
OpenShift Container Platform/{product-title}

jc-berger replied:
When I write `{product-title}` and then go view the preview, `{product-title}` does not render `OpenShift Container Platform`

Preeticp commented on modules/about-redhat-openshift-gitops.adoc:
```diff
@@ -0,0 +1,26 @@
+
+* Ensure that the clusters have similar states for configuration, monitoring, and storage
+* Recover or recreate clusters from a known state
+* Apply or revert configuration changes to multiple OpenShift Container Platform clusters
```
OpenShift Container Platform/{product-title}

jc-berger replied:
See comment above. May I keep it the way it is, fully writing out `OpenShift Container Platform`?

Preeticp commented:
Looking good @jc-berger  Some minor nits. Please fix them and ask the OCP team for a review and merge. Thank you!

Preeticp commented:
At the top of this PR, in the PR description box please replace the existing content with:

- Adds an overview for GitOps and addresses Jira Issue, #2475.
- Targeted to OCP 4.7
- Aligned team label: Dev Tools
- Acked by SME/QE

bigkevmcd commented on modules/about-gitops.adoc:
```diff
@@ -0,0 +1,12 @@
+
+GitOps is a declarative way to implement continuous deployment for cloud native applications. You can use GitOps to create repeatable processes for managing OpenShift Container Platform clusters and applications across multi-cluster Kubernetes environments. GitOps handles and automates complex deployments at a fast pace, saving time during deployment and release cycles. Its workflow pushes an app ...
+
+GitOps is a set of practices that use Git pull requests to manage infrastructure and application configurations. Its Git repository is the only source of truth for system and application configuration. This Git repository contains a declarative description of the infrastructure you need in your specified environment and contains an automated process to make your environment match the described st ...
```
Should probably call the set of practices out explicitly here.

bigkevmcd commented on modules/about-redhat-openshift-gitops.adoc:
```diff
@@ -0,0 +1,26 @@
+[id="about-redhat-openshift-gitops_{context}"]
+= About {gitops-title}
+
+{gitops-title} ensures consistency in applications when you deploy them to different clusters in different environments, like: development, stage, and production. {gitops-title} organizes the deployment process around the code repositories and makes them the central element. It always has at least two repositories:
```
```suggestion
{gitops-title} ensures consistency in applications when you deploy them to different clusters in different environments, like: development, stage, and production. {gitops-title} organizes the deployment process around configuration repositories and makes them the central element. It always has at least two repositories:
```

bigkevmcd reviewed (commented):
I suspect we need to include the expanded definition that includes manual deployments too.

adellape commented on modules/about-redhat-openshift-gitops.adoc:
```diff
@@ -0,0 +1,26 @@
+[id="about-redhat-openshift-gitops_{context}"]
+= About {gitops-title}
+
+{gitops-title} ensures consistency in applications when you deploy them to different clusters in different environments, like: development, staging, and production. {gitops-title} organizes the deployment process around the configuration repositories and makes them the central element. It always has at least two repositories:
```
s/like:/such as/

adellape commented on modules/about-redhat-openshift-gitops.adoc:
```diff
@@ -0,0 +1,26 @@
+
+These repositories contain a declarative description of the infrastructure you need in your specified environment. They also contain an automated process to make your environment match the described state.
+
+{gitops-title} uses Argo CD to maintain cluster resources. Argo CD is an open-source declarative tool for the continuous integration and continuous deployment (CI/CD) of applications. {gitops-title} implements Argo CD as a controller so that it continuously monitors application definitions and configurations defined in a Git repository. Then, Argo CD compares the specified state of these configur ...
```
I assume the spelling of "Argo CD" here (with a space) is correct, though I do see in https://docs.openshift.com/container-platform/4.6/architecture/argocd.html it's spelled (probably incorrectly) mashed-up as "ArgoCD".

adellape replied:
This paragraph ends up kinda long, maybe consider breaking it up either starting at "{gitops-title} implements Argo CD" or "These reports allow"?

adellape commented on modules/about-redhat-openshift-gitops.adoc:
```diff
@@ -0,0 +1,26 @@
+[id="key-features_{context}"]
+== Key features
+
+{gitops-title} helps you:
```
Per IBMSG, avoid continuing the lead-in sentence after the list. So maybe something like "Red Hat OpenShift GitOps helps you automate the following tasks:"

adellape commented on cicd/gitops/understanding-openshift-gitops.adoc:
```diff
@@ -0,0 +1,12 @@
+[id="understanding-openshift-gitops"]
+= Understanding OpenShift GitOps
```
I see there's sort of three different names going on here, between "OpenShift GitOps", plain ol' "GitOps", and "Red Hat OpenShift GitOps" (which I see is set as `:gitops-title:`).

If the "About GitOps" section is intended to just discuss the general purpose of GitOps overall (not specific to the RH product), maybe re-title that one as something like "About GitOps principles" or "About GitOps methodology"?

And if "OpenShift GitOps" is not an approved shortened name (I see `:gitops-shortname:` is just set to `GitOps` below), probably should use "Red Hat OpenShift GitOps" in the title here and the left-nav (in `_topic_map.yml`).

adellape reviewed (commented):
Looks cool! Mostly suggestions inline.

Also remember to:

* Remove "WIP" from your PR title.
* Trim down your commit message to just what you want to appear in the final merge. Right now it's many lines long after your recent squash. Could use `git commit --amend` to just keep the first line ("created understanding gitops concept module"); any line that doesn't start with a `#` will appear in the message.

jc-berger commented:
Thanks, @adellape! I addressed your feedback, thanks so much! I added a `#` to my numerous commit messages, but it still appears long, so I must have missed something, lol. 

Let me know if you think it looks good to merge from a doc perspective, thanks!

jeana-redhat reviewed (approved):
approving exception

jeana-redhat commented:
/cherrypick enterprise-4.7
